### ISSUE TITLE
Support Multiple Pages in Edit Entry

### ISSUE THREAD

**Issue by zackkatz** [gravityview/GravityView#718]
- [ ] Add support for Review Page: https://www.gravityhelp.com/documentation/article/gform_review_page/
- [ ] Support Conditional Logic page breaks
- [ ] Make sure validation is properly handled for multi-page forms

**Comment by zackkatz** [comment]
@soulseekah And this is another :-)

**Comment by soulseekah** [comment]
I think this should be left for EE 2.0 (the rewrite)?

Either way, since edited form fields are very flexible, where do we add the page breaks to? How do we make them configurable? Conditional pagination is even harder.

Our current field editor is very limited. I'm trying to think hard on how to make things work in a way that would be intuitive UX wise and would make sense in the backend wise as well.

For the backend we need to know:

1. Whether a page was added to the edit context.
2. Which of the fields in that specific page are being added.
3. Can we even move fields from one page to the other?
4. Do we preserve conditional logic for pages, or do we just allow pagination unrelated to the pages in the actual form?
5. Do we reset values for pages that become disabled due to conditional logic?

There's a lot of fundamental discussion there about this needed here, as outlined above. We should discuss this in a lot of depth.

**Comment by zackkatz** [comment]
@soulseekah I think we landed on only supporting multiple pages when showing the whole form. Otherwise we are recreating Gravity Forms!

**Comment by soulseekah** [comment]
Makes sense. Should we be saving the values each and every step or only on the last submission step? Right now after some of my experiments it's saving them on every step of the way.

**Comment by soulseekah** [comment]
@zackkatz can you respond above, please?

**Comment by zackkatz** [comment]
Since it seems easier and I don’t have a firm opinion either way, let’s do each step for now, and then people can complain if that’s not what they want. @soulseekah

@rafaehlers @jgrisham what do you think?

**Comment by rafaehlers** [comment]
It also seems "safer" saving it page by page.

**Comment by soulseekah** [comment]
Note that Gravity Forms does not save the form page by page :)

I'll implement

**Comment by soulseekah** [comment]
Highly experimental.

Add `add_filter( 'gravityview/features/paged-edit', '__return_true' );` to activate. Make sure you don't have fields setup in the Edit configuration.

If it behaves well overall I can start writing unit tests for this.
Otherwise needs testing.

**Comment by vendojo** [comment]
@soulseekah just tried this feature and while it correctly showed the steps at the top of the form, it would finish the editing process right after clicking the 'Next' button. So right now, the 'Next' button is not loading the next page of this form but already submitting the edited entry.

Let me know if you need any further information.

**Comment by soulseekah** [comment]
@vendojo Thank you for testing. This is a limitation, the entry fields are submitted on every single page right now. But the pages should still be advancing correctly and showing the specific page fields. Is this not what you're seeing? Can you show a video?

**Comment by vendojo** [comment]
Here's the video: https://www.loom.com/share/101bbc772e994b75842f02413b86f72f

There seems to be an issue with the form's confirmation type. I was using a redirect to another page on submission (see ![screenshot](https://user-images.githubusercontent.com/56757832/67634407-dead8f80-f8bb-11e9-9b4b-1b12811bbde1.png)). As the entry is submitted on every page, it does apparently also trigger this confirmation behavior. I switched to a simple 'text' confirmation and it advanced correctly.

However, the form also advances to the next page when clicking the 'Previous' button (see 
![screenshot](https://user-images.githubusercontent.com/56757832/67634441-32b87400-f8bc-11e9-926b-3dc7d2f71039.png)). It would be better if it would actually go back one page when clicking this button, right?

**Comment by soulseekah** [comment]
Issues found:

- [ ] Previous skips 2 pages when clicked, going in a loop.
- [ ] Redirecting After Edit setting is pulled on every page.

Redirecting to another page via the Form Settings is not being triggered for me, but:

- [ ] Check and make sure we're proactively suppressing all confirmations and notifications on our edit pages.

**Comment by soulseekah** [comment]
@vendojo please test the latest develop.zip please?

**Comment by vendojo** [comment]
> @vendojo please test the latest develop.zip please?

That's awesome, thanks mate! Works great, now, even when I have the 'Redirect to another page' setting enabled.

Just two minor things:

Ideally, the 'previous' and 'next' buttons would inherit the CSS classes and translation defined by the Gravity Forms plugin. That way you wouldn't have to translate/style them twice:
![screenshot-bwe6id myraidbox de-2019 10 29-10_20_10](https://user-images.githubusercontent.com/56757832/67753970-e42ee500-fa35-11e9-9274-0d22a7cd1fe4.jpg)

There appears a 'Entry updated' notification now. Don't know if this was built in intentionally or if it's a side effect. I just noticed it in the new develop version.
![screenshot-bwe6id myraidbox de-2019 10 29-10_17_11](https://user-images.githubusercontent.com/56757832/67754330-9070cb80-fa36-11e9-8403-73e383926615.jpg)

**Comment by soulseekah** [comment]
@vendojo you should target the CSS correctly.

Regarding the update message it was there but with a link. It's without a link now. You can filter the message  by hooking into the new 'gravityview/edit_entry/page/success` filter.

**Comment by vendojo** [comment]
@soulseekah fair enough. Seems to be working fine. When will this feature be available in the stable version?

**Comment by soulseekah** [comment]
By the end of November, we hope. We need more testing before rolling it out.

**Comment by zackkatz** [comment]
@soulseekah 

- Validation works
- I think we should make it clear this _won't work with Review Page_ because we're saving at each page

### "Support Conditional Logic page breaks" still isn't working.

Here's a two-page form with "page conditional logic" https://i.gravityview.co/NaxtkU

- Create an entry where the checkbox is checked and number filled out
- Edit that entry in a View
- Uncheck the box, click next page
- What happens: It still shows the next page, with the number value cleared
- What should happen: It does not show the next page; it shows the last page of the form with "Entry Updated" message

**Comment by soulseekah** [comment]
On a related note, the " Enable Next Button Conditional Logic " feature will very very probably not work either. I'll test after I manage to fix the conditional logic page breaks.

**Comment by soulseekah** [comment]
Haven't been able to get the Next Button Logic to work yet, but Conditional hide/shows of pages has been implemented and briefly tested. Please try it out whoever is interested.

**Comment by zackkatz** [comment]
@soulseekah Can you share details on what the blocker is for the Next Button Conditional Logic functionality?

It'd be nice to have full feature parity.

**Comment by soulseekah** [comment]
Next Button Logic involves pretty complicated logic code that I wasn't able to call correctly. The next option would be to rewrite the thing (this is pure JavaScript). On the other hand it's such a non-critical feature that it wouldn't be missed, I'm sure. Can we wait for someone to ask for it?

**Comment by rafaehlers** [comment]
https://secure.helpscout.net/conversation/1045476411/24343

**Comment by soulseekah** [comment]
> https://secure.helpscout.net/conversation/1045476411/24343

Initial support landed in 2.5.1 but not documented.

It can be turned on using `add_filter( 'gravityview/features/paged-edit', '__return_true' );` for testing by customers who are willing to do so.

**Comment by bennemann** [comment]
https://secure.helpscout.net/conversation/1315040232/29867

**Comment by bennemann** [comment]
https://secure.helpscout.net/conversation/1328541416/30112/

**Comment by bennemann** [comment]
https://secure.helpscout.net/conversation/2163809395/47223/

**Comment by bennemann** [comment]
https://secure.helpscout.net/conversation/2228455552/48645

**Comment by rafaehlers** [comment]
https://secure.helpscout.net/conversation/2231665311/48710

**Comment by rafaehlers** [comment]
https://secure.helpscout.net/conversation/2728285946/57844 mentions that the `add_filter( 'gravityview/features/paged-edit', '__return_true' );` filter breaks this other filter used to customize messages: `add_filter( 'gravityview/edit_entry/success', 'gv_edit_entry_success', 10, 4 );`

**Comment by rafaehlers** [comment]
https://secure.helpscout.net/conversation/2872279116/61959

**Comment by rafaehlers** [comment]
https://secure.helpscout.net/conversation/2882495259/62207
